File: status/statuscalc.R
# HYDROSOS LAKE VICTORIA BASIN WORKSHOP
# script to calculate hydrological status from flow data
# Katie Facer-Childs 15-08-2023

### PARTICIPANTS WILL NEED TO EDIT LINES 14,15,16, (possibly 27 & 79), 88,89

# set working directory (you'll need to replace \ with / if you copy the path)
#setwd("./")

################################# PREPARATION #################################

# read in catchment list
#stations<-read.csv("stations.csv")
stationid<-39001
input_directory<-"./example_data/input/"
output_directory<-"./example_data/output_R/"

# read in flow data
#flowdata<-read.csv(paste0(stationid,".csv"))
flowdata<-read.csv(sprintf("%s%s.csv", input_directory,stationid))           

#set the column names
colnames(flowdata)<-c("date","flow")
# format the date column (this might cause difficulties, please look at the 
# date as it appears in the column and amend the format accordingly 
# e.g. 1991-02-30 would be "%Y-%m-%d")
flowdata$date<-as.Date(flowdata$date,format="%d/%m/%Y")

### check the timeseries is complete (sequential) - we need missing data to be
### set as NAs, not missing from the dataset

if (length(flowdata$date)!=as.numeric(max(flowdata$date)-min(flowdata$date))){
  fulldata<-data.frame(date=c(as.Date(seq(min(flowdata$date),max(flowdata$date),by=1),format="%Y-%m-%d")))
  flowdata<-merge(fulldata,flowdata,by="date",all.x=T)
}

flowdata$date<-as.Date(flowdata$date)

#make a month and a year column
flowdata$month<-as.numeric(format(flowdata$date,format="%m"))
flowdata$year<-as.numeric(format(flowdata$date,format="%Y"))

# plot the daily timeseries
plot(flowdata$date,flowdata$flow, type="l",xlab="Date",ylab="flow (cumecs)")

# is the data long enough to calculate a Long Term Average?
startYear<-min(flowdata$year)
endYear<-max(flowdata$year)
print(paste0("there are ",endYear-startYear," years of data in this file"))
# how much is missing data?
navals<-sum(is.na(flowdata$flow))
flowlength<-length(flowdata$flow)
print(paste0("there are ",navals, " missing values, which is ", 
             sprintf((navals/flowlength)*100,fmt='%#.2f'), "% missing")) 

# is there too much missing data for any particular month? 

# this counts the non-NA values in each month-year, 
# divides it by the length of the month, 
# and multiplies it by 100 to get the percent complete

# we will use this to set months with less than 50% completeness to NA
flowcompleteness<-aggregate(flow ~ month + year, flowdata, 
                            FUN=function(x) ((sum(!is.na(x)))/(length(x)))*100,
                            na.action=NULL)
colnames(flowcompleteness)<-c("month","year","pc_complete")


## FUTURE STEP - INFILLING MISSING DATA

##################### STEP 1 CALCULATE MONTHLY MEAN FLOWS #####################

mon_flow<-aggregate(flow ~ month + year, flowdata,
                    FUN=mean, na.rm=TRUE, na.action=NULL)
# set those NA values where the months are too incomplete
mon_flow$flow[which(flowcompleteness$pc_complete<50)]<-NaN

# plot monthly data
mon_flow$date<-as.Date(paste0("01-",mon_flow$month,"-",mon_flow$year),
                       format="%d-%m-%Y")
plot(mon_flow$date,mon_flow$flow, type="l",xlab="Date",ylab="flow (cumecs)")

######## STEP 2 CALCULATE MONTHLY MEAN FLOWS AS A PERCENT OF AVERAGE #########

# here we will use 1991-2020, you will need to edit these dates if this is not
# suitable for your data

stdStart<-1991
stdEnd<-2020

# again let's do some completeness checks

LTAcheck<-aggregate(flow ~ month, 
                    mon_flow[which(mon_flow$year==stdStart & mon_flow$month==1):
                             which(mon_flow$year==stdEnd & mon_flow$month==12),],
                    FUN=function(x) sum(!is.na(x)),
                    na.action=NULL)
print(LTAcheck)

# how close to 30 years do you have for each month of the year?!

# calculate the long term average (LTA) for each calendar month
LTAdata<-aggregate(flow ~ month, 
                   mon_flow[which(mon_flow$year==stdStart & mon_flow$month==1):
                              which(mon_flow$year==stdEnd & mon_flow$month==12),],
                   FUN=mean, na.rm=TRUE, na.action=NULL)


# use this as a look up to calculate each month as a percent of average
# (this is called a "for loop")
mon_flow$flowperc<-NA

flowpercfun<-  function(x) (mon_flow$flow[x]/LTAdata$flow[which(LTAdata$month==mon_flow$month[x])])*100
mon_flow$flowperc<-unlist(lapply(1:length(mon_flow$flow),FUN=flowpercfun))

# another plot check
plot(mon_flow$date,mon_flow$flowperc, type="l",xlab="Date",ylab="Flow as % average")


###################### STEP 3 CALCULATE RANK PERCENTILES ######################

# rank percentile uses the Weibull distribution to rank the month 
# against the historic data
# i/N+1 (where i is the rank of the current month and N is the number of months 
# in the period of record)

mon_flow$rank<-NA
mon_flow$rankperc<-NA

for (imonth in 1:12){
  mon_flow$rank[which(mon_flow$month==imonth)]<-rank(mon_flow$flowperc[which(mon_flow$month==imonth)], na.last='keep')
}

for (imonth in 1:12){
  mon_flow$rankperc[which(mon_flow$month==imonth)]<-mon_flow$rank[which(mon_flow$month==imonth)]/(sum(!is.na(mon_flow$flow[which(mon_flow$month==imonth)]))+1)
}

# set the NA values as rank NA
mon_flow$rank[is.na(mon_flow$flow)]<-NaN
mon_flow$rankperc[is.na(mon_flow$flow)]<-NaN

# plot again
plot(mon_flow$date,mon_flow$rankperc, type="l",xlab="Date",ylab="Rank Percentile")


####################### STEP 4 ASSIGN STATUS CATEGORIES #######################

# set up percentile categories
# NOTE THESE ARE LIKELY TO BE CHANGED TO BE MORE "STANDARD" SOON
pc_cat<-data.frame(
  category=c("low","below_normal","normal","above_normal","high"),
  lowerlim=c(0,0.13001,0.28001,0.72,0.87),
  upperlim=c(0.13,0.28,0.71999,0.86999,1),
  hexcolour=c("#CD233f","#FFA883","#E7E2BC","#8ECEEE","#2C7DCD")
)

catlookup <- function (x) ifelse(is.na(mon_flow$rankperc[x]),
                                 NaN,
                                 which(mon_flow$rankperc[x] >= pc_cat$lowerlim & mon_flow$rankperc[x] <= pc_cat$upperlim))


mon_flow$flowcat<-unlist(lapply(1:length(mon_flow$flow),FUN=catlookup))
mon_flow$catname<-pc_cat[mon_flow$flowcat,1]
# print the last few values
print(tail(mon_flow$catname))

mon_flow$hexcolour<-as.character(pc_cat[mon_flow$flowcat,4])
barplot(tail(mon_flow$flowperc) ~ tail(mon_flow$date),col=(tail(mon_flow$hexcolour)),xlab="Date",ylab="Flow Percentile")


##################### STEP 5 WRITE OUT DATA FOR THE PORTAL #####################

outData<-data.frame(
  date=as.Date(mon_flow$date,format="%Y-%m-%d"),
  flowcat=mon_flow$flowcat)
write.csv(outData,sprintf("%scat_%s.csv", output_directory,stationid),row.names=F)

File: status/statuscalc.py
"""
BASED ON STATUSCALCV2.R BY KATIE FACER-CHILDS
EZRA KITSON  25-09-2025
"""

import argparse
import pandas as pd
import numpy as np
import os
from pathlib import Path

parser = argparse.ArgumentParser(
                    prog='StatusCalc v3 PYTHON',
                    description='Calculates status based on daily timeseries for the HydroSOS portal',
                    epilog='Katie F-C, Ezra K, UKCEH, 14052024')


parser.add_argument('input_directory', help='input directory, should ONLY contain .csv daily timeseries, see GitHub for examples.')        
parser.add_argument('output_directory', help='directory files will be saved to as cat_{input_file}.csv')     
parser.add_argument('--startYear', help='start of the year range that will be used to calculate the reference average.')
parser.add_argument('--endYear', help='end of the year range that will be used to calculate the reference average.')
parser.add_argument('--debugging', help='print debugging')

args = parser.parse_args()

if args.startYear:
    stdStart=int(args.startYear)
else: 
    print("No start year set, defaulting to 1991.")
    stdStart=1991

if args.endYear:
    stdEnd=int(args.endYear)
else: 
    print("No end year set, defaulting to 2020.")
    stdEnd=2020

assert stdStart < stdEnd, "startYear must be greater than endYear"

#stationid="39001"
#input_directory="./example_data/input/"
#output_directory="./example_data/output_Python/"

Path(args.output_directory).mkdir(parents=True, exist_ok=True)
Path(f"{args.output_directory}/statusBands").mkdir(parents=True, exist_ok=True)

for f in os.listdir(args.input_directory):

    if f.endswith('.csv'):
        print(f)
        flowdata = pd.read_csv(f"{args.input_directory}{f}")
        flowdata.columns = ['date','flow']
        flowdata['date'] = pd.to_datetime(flowdata['date'], format="%d/%m/%Y")

        #check dates are sequential
        diff = pd.date_range(start = flowdata['date'].min(), end = flowdata['date'].max() ).difference(flowdata['date'])
        if len(diff) > 0:
            flowdata.set_index('date', inplace=True)
            for md in diff: 
                flowdata.loc[md,'flow'] = pd.NA
            flowdata.reset_index(inplace=True)

        #month and year column
        flowdata['month'] = flowdata['date'].dt.month
        flowdata['year'] = flowdata['date'].dt.year

        #check whether or not there is enough data? 
        print(f"There are {flowdata['year'].max() - flowdata['year'].min()} years of data in this file.")
        print(f"There are {sum(flowdata['flow'].isnull())} missing data points, which is {np.round(sum(flowdata['flow'].isnull())/len(flowdata) * 100,4)}% of the total data")

        """ STEP 1: CALCULATE MEAN MONTHLY FLOWS """

        #calculate percentage completeness for each year/month
        groupBy = (flowdata.groupby(['month','year']).count()['flow']/flowdata.groupby(['month','year']).count()['date']) * 100
        groupBy = pd.DataFrame(groupBy)
        groupBy.rename(columns={0:'monthly%'}, inplace=True)
        #calculate mean flows for each year/month
        groupBy['mean_flow'] = flowdata.groupby(['month','year'])['flow'].mean()
        #set the mean flow to NAN if there is less than 50 % data
        groupBy.loc[groupBy['monthly%'] < 50,'mean_flow'] = pd.NA
        groupBy.reset_index(inplace=True)

        """ STEP 2: CALCULATE MEAN MONTHLY FLOWS AS A PERCENTAGE OF AVERAGE REFERENCE PERIOD """

        #calculate long term average
        LTA = groupBy[(groupBy['year'] >= stdStart) & (groupBy['year'] <= stdEnd)].groupby(['month'])['mean_flow'].mean()

        #divide each month by this long term average
        skip_file = False
        for i in range(1,13):
            if i not in LTA.index:
                print("ERROR: Month %i missing in Long Term Average for file %s. Skipping file" % (i,f))
                skip_file = True
                break
            groupBy.loc[groupBy['month'] == i,'percentile_flow'] = groupBy['mean_flow'][groupBy['month'] == i]/LTA[i] * 100

        if skip_file:
            continue

        """ STEP 3: CALCULATE RANK PERCENTILES OF REFERENCE PERIOD """

        refBy = groupBy[(groupBy['year'] >= stdStart) & (groupBy['year'] <= stdEnd)]

        # na values automatically set as rank na
        for i in range(1,13):
            refBy.loc[refBy['month'] == i, 'weibell_rank'] = refBy.loc[refBy['month'] == i, 'percentile_flow'].rank(na_option='keep')/(refBy.loc[refBy['month'] == i, 'percentile_flow'].count()+1)

        targetRanks = {0.10,0.25,0.75,0.9}
        thresholdDict = {}

        for i in range(1,13):
            ranks = np.array(refBy.loc[refBy['month'] == i, 'weibell_rank'])
            percentiles = np.array(refBy.loc[refBy['month'] == i, 'percentile_flow'])
            for j in targetRanks:
                #find the closest rank to the target ranks above and below
                lower_vals = ranks[ranks <= j]
                higher_vals = ranks[ranks >= j]
                closest_higher = np.min(higher_vals) if higher_vals.size > 0 else False
                closest_lower = np.max(lower_vals) if lower_vals.size > 0 else False
                closest_higher_idx = np.where(ranks == closest_higher)[0][0] if closest_higher else False
                closest_lower_idx = np.where(ranks == closest_lower)[0][0] if closest_lower else False
                #find the percentile values matching to the closet rank
                higher_percentile = percentiles[closest_higher_idx] if closest_higher_idx else False
                lower_percentile = percentiles[closest_lower_idx] if closest_lower_idx else False
                #use a linear interpolation to get the percentile value of the target rank based on these two values
                # this occurs if the target rank perfectly matches an observed rank
                if higher_percentile == lower_percentile:
                    interpolated_percentile = lower_percentile
                # this occurs if no observed ranks were higher than the target rank, in which case set as the lower percentile
                elif higher_percentile == False:
                    interpolated_percentile = lower_percentile
                # this occurs if no observed ranks were lower than the target rank, in which case set as the higher percentile
                elif lower_percentile == False:
                    interpolated_percentile = higher_percentile
                # otherwise linearly interpolate the percentile value from the closest higher and lower 
                else:
                    interpolated_percentile = lower_percentile + ((j - closest_lower) / (closest_higher - closest_lower)) * (higher_percentile - lower_percentile) 
                if args.debugging: 
                    print(f"Month : {i}")
                    print(f"Percentiles in ref: {percentiles}")
                    print(f"Ranks in ref: {ranks}")
                    print(f"Target rank: {j}")
                    print(f"Closest lower rank: {closest_lower}")
                    print(f"Closest  higher rank: {closest_higher}")
                    print(f"Closest lower percentile: {lower_percentile}")
                    print(f"Closest higher percentile: {higher_percentile}")
                    print(f"Interpolated percentile: {interpolated_percentile}")
                #add this to the threshold dictionary
                thresholdDict[i,j] = interpolated_percentile
                #add min, mean and max to thresholdDict to 
                thresholdDict[i,'max'] = np.nanmax(percentiles)
                thresholdDict[i,'median'] = np.nanmedian(percentiles)
                thresholdDict[i,'min'] = np.nanmin(percentiles)

        """ STEP 4: ASSIGN STATUS CATEGORIES """

        def flow_status(percentile, month, thresholdDict):
            status = pd.NA
            if percentile <= thresholdDict[month,0.1]:
                status = 1
            elif percentile <= thresholdDict[month,0.25]:
                status = 2
            elif percentile <= thresholdDict[month,0.75]:
                status = 3
            elif percentile <= thresholdDict[month,0.9]:
                status = 4
            elif percentile > thresholdDict[month,0.9]:
                status = 5
            return status

        for i in groupBy.index:
            groupBy.loc[i,'category'] = flow_status(percentile=groupBy.loc[i,'percentile_flow'],month=groupBy.loc[i,'month'],thresholdDict=thresholdDict)


        """ STEP 5: WRITE DATA """
        groupBy['date'] = pd.to_datetime(groupBy[['year', 'month']].assign(DAY=1))
        groupBy['date'] = groupBy['date'].dt.strftime('%Y-%m-%d')
        groupBy['category'] = groupBy['category'].astype('Int64')
        groupBy.sort_values(['year','month']).filter(['date','category']).to_csv(f"{args.output_directory}cat_{f}", index=False)
        forecastBands = pd.DataFrame.from_dict(pd.Series(thresholdDict).unstack())
        forecastBands.to_csv(f"{args.output_directory}/statusBands/{f.split('.')[0]}_bands.csv")
       




        
